ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Client.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Ouroboros.Network.Protocol.BlockFetch.Client where
+
+import Protocol.Core
+import Pipes (Producer)
+import qualified Pipes
+
+import Ouroboros.Network.Protocol.BlockFetch.Type
+
+{-------------------------------------------------------------------------------
+  Client stream of @'BlockFetchClientProtocol'@ protocol
+-------------------------------------------------------------------------------}
+
+data BlockFetchClientSender range m a where
```

Comment:
This type ought to be

```Haskell
newtype BlockFetchClientSender range m a = BlockFetchClientSender
  { runBlockFetchClientSender :: m (BlockFetchClientRequest range m a)
  }

data BlockFetchClientRequest range m a where
  BlockFetchClientRange :: range -> BlockFetchClientSender range m a -> BlockFetchClientRequest range m a
  BlockFetchClientDone :: a -> BlockFetchClientRequest range m a
```

This will make `blockFetchClientSenderFromProducer` and `blockFetchClientSenderStream` more pleasant and easy to use: you can have

```Haskell
-- No 'm' out front of the rightmost type.
blockFetchClientSenderFromProducer :: Monad m => Producer range m a -> BlockFetchClientSender range m a

-- blockFetchClientSenderStream . blockFetchClientSenderFromProducer
-- is well-typed under this presentation.
```

Reply:
You're right, indeed this will hidie the monad `m`.